File: ts/2025/day09.ts
type Point = [x: number, y: number]

const minmax = (a: number, b: number): [number, number] => {
	return [Math.min(a, b), Math.max(a, b)]
}

const cache = new Map<string, boolean>()

const contains = (pts: Point[], [x, y]: Point): boolean => {
	const key = `${x},${y}`
	let ret = cache.get(key)
	if (ret !== undefined) { return ret }

	let cnt = 0
	for (let i = 0; i < pts.length; i++) {
		const [[ax, ay], [bx, by]] = [pts[i]!, pts[(i + 1) % pts.length]!]
		const [x1, x2] = minmax(ax, bx)
		const [y1, y2] = minmax(ay, by)
		if (x === x1 && x === x2 && y >= y1 && y <= y2) {
			cache.set(key, true)
			return true
		}
		if (y === y1 && y === y2 && x >= x1 && x <= x2) {
			cache.set(key, true)
			return true
		}
		if ((y1 > y) !== (y2 > y)) {
			if (x < x1 + (x2 - x1) * (y - y1) / (y2 - y1)) {
				cnt++
			}
		}
	}
	ret = cnt % 2 === 1
	cache.set(key, ret)
	return ret

}

const intersect = (pts: Point[], a: Point, b: Point): boolean => {
	const [ax, ay] = a
	const [bx, by] = b

	const key = `${ax},${ay}-${bx},${by}`
	let ret = cache.get(key)
	if (ret !== undefined) { return ret }

	for (const [i, [cx, cy]] of pts.entries()) {
		const [dx, dy] = pts[(i + 1) % pts.length]!
		const [x1, x2] = minmax(cx, dx)
		const [y1, y2] = minmax(cy, dy)
		if (x1 === x2) {
			if (ax < x1 && x1 < bx && ay < y2 && y1 < by) {
				cache.set(key, true)
				return true
			}
		} else {
			if (ay < y1 && y1 < by && ax < x2 && x1 < bx) {
				cache.set(key, true)
				return true
			}
		}
	}
	cache.set(key, false)
	return false
}

let p1 = 0, p2 = 0
const lines: Point[] = (await Bun.stdin.text()).trim().split("\n")
	.map(line => line.split(",").map(Number))
	.map(([x = 0, y = 0]) => [x, y])
for (const [i, [ax, ay]] of lines.slice(0, -2).entries()) {
	for (const [_, [bx, by]] of lines.slice(i + 1).entries()) {
		const [x1, x2] = minmax(ax, bx)
		const [y1, y2] = minmax(ay, by)

		if (x1 === x2 || y1 === y2) { continue }

		const area = (x2 - x1 + 1) * (y2 - y1 + 1)
		p1 = Math.max(p1, area)

		if (area <= p2) { continue }

		if (contains(lines, [x1, y1]) &&
			contains(lines, [x1, y2]) &&
			contains(lines, [x2, y1]) &&
			contains(lines, [x2, y2]) &&
			!intersect(lines, [x1, y1], [x2, y2])) {
			p2 = Math.max(p2, area)
		}
	}
}
console.log("Part 1: %d", p1)
console.log("Part 2: %d", p2)

File: py/2025/day09.py
import sys
from functools import cache
from itertools import combinations, chain


p1, p2 = 0, 0
D = sys.stdin.read()
lines = [list(map(int, line.split(","))) for line in D.split("\n") if line]


@cache
def valid(x, y) -> bool:
    cnt = 0
    for i, (x1, y1) in enumerate(lines):
        x2, y2 = lines[(i + 1) % len(lines)]
        x1, x2 = [x1, x2] if x1 <= x2 else [x2, x1]
        y1, y2 = [y1, y2] if y1 <= y2 else [y2, y1]
        if x == x1 and x == x2 and y1 <= y <= y2:
            return True
        if y == y1 and y == y2 and x1 <= x <= x2:
            return True
        if (y1 > y) != (y2 > y):
            if x < x1 + (x2 - x1) * (y - y1) / (y2 - y1):
                cnt += 1
    return (cnt % 2) == 1


@cache
def intersect(a, b) -> bool:
    ax, ay = a
    bx, by = b
    for i, (x1, y1) in enumerate(lines):
        x2, y2 = lines[(i + 1) % len(lines)]
        x1, x2 = [x1, x2] if x1 <= x2 else [x2, x1]
        y1, y2 = [y1, y2] if y1 <= y2 else [y2, y1]
        if x1 == x2:
            if ax < x1 < bx and ay < y2 and y1 < by:
                return True
        else:
            if ay < y1 < by and ax < x2 and x1 < bx:
                return True
    return False


for a, b in combinations(lines, 2):
    x1, x2, y1, y2 = list(chain(*list(zip(a, b))))
    x1, x2 = [x1, x2] if x1 <= x2 else [x2, x1]
    y1, y2 = [y1, y2] if y1 <= y2 else [y2, y1]
    if x1 == x2 or y1 == y2:
        continue
    area = (x2 - x1 + 1) * (y2 - y1 + 1)
    p1 = max(area, p1)

    if area <= p2:
        continue
    if (
        valid(x1, y1)
        and valid(x1, y2)
        and valid(x2, y1)
        and valid(x2, y2)
        and not intersect((x1, y1), (x2, y2))
    ):
        p2 = max(area, p2)

print(f"The result for part 1: {p1}")
print(f"The result for part 2: {p2}")
